src/08-recommendations/03-pipeline-recommend.ts

```diff
@@ -0,0 +1,325 @@
+/*
+ * This variation tries the same technique of feeding candidates to the LLM to have it
+ * choose and justify recommendations, but it differs from the previous attempts in two ways:
+ *
+ * - it sources the artwork candidates from our actual recommendations pipeline via a tool call
+ * - it provides fewer candidates
+ *
+ * Because this makes a real query to Metaphysics for artwork recommendations, it requires
+ * a valid user ID and access token to be provided as env vars / in the .env file.
+ */
+
+import OpenAI from "openai"
+import dotenv from "dotenv"
+import chalk from "chalk"
+
+dotenv.config()
+const openai = new OpenAI()
+
+const userID = process.env.RECOMMENDATIONS_USER_ID
+const accessToken = process.env.RECOMMENDATIONS_USER_ACCESS_TOKEN
+
+if (!userID || !accessToken) {
+  throw new Error(
+    "Please provide a RECOMMENDATIONS_USER_ID and corresponding RECOMMENDATIONS_USER_ACCESS_TOKEN in the .env file to run this experiment"
+  )
+}
+
+// Settings
+
+/** fetch this many candidates from each of the recommendation fields */
+const NUM_CANDIDATES_PER_FIELD = 10
+
+/** applies to the `artworksForUser` field */
+const INCLUDE_BACKFILL = true
```

Comment:
Question: I see this is eventually passed on to MP, but I'm not clear on what it actually does. Generate responses in realtime, if there are none?

Reply:
Ah, now see the [answer](https://artsy.slack.com/archives/C033E79AQG4/p1715876003763329?thread_ts=1715869812.362219&cid=C033E79AQG4) in the slack message linked to below.